File: 2-add-two-numbers/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode()
        
        node = head
        carry = 0
        
        while l1 and l2:
            value = l1.val + l2.val + carry
            carry = 1 if value > 9 else 0
            node.next = ListNode(value % 10)
            node, l1, l2 = node.next, l1.next, l2.next
        
        lst = l1 if l1 else l2
        
        while lst and carry:
            value = lst.val + carry
            carry = 1 if value > 9 else 0
            node.next = ListNode(value % 10)
            node, lst = node.next, lst.next
        
        if carry:
            node.next = ListNode(1)
        if lst:
            node.next = lst
            
        return head.next

File: 2-add-two-numbers/2-add-two-numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func addWithCarry(val1, val2 int) (val, carry int) {
    val = val1 + val2
    if val > 9 {
        carry = 1
    }
    val %= 10
    return
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    head := &ListNode{}
    
    node := head
    val, carry := 0, 0
    
    for l1 != nil && l2 != nil {
        val, carry = addWithCarry(l1.Val, l2.Val + carry)
        node.Next = &ListNode{Val: val}
        node, l1, l2 = node.Next, l1.Next, l2.Next
    }
    
    var lst *ListNode
    if l1 != nil {
        lst = l1
    } else {
        lst = l2
    }
    
    for lst != nil && carry != 0 {
        val, carry = addWithCarry(lst.Val, carry)
        node.Next = &ListNode{Val: val}
        node, lst = node.Next, lst.Next
    }
    
    if carry == 1 {
        node.Next = &ListNode{Val: 1}
    }
    if lst != nil {
        node.Next = lst
    }
    
    return head.Next
}
